synapse/handlers/message.py

```diff
@@ -886,10 +885,21 @@ async def create_and_send_nonmember_event(
                 event.sender,
             )
 
-            spam_check = await self.spam_checker.check_event_for_spam(event)
-            if spam_check is not synapse.spam_checker_api.Allow.ALLOW:
+            spam_check_result = await self.spam_checker.check_event_for_spam(event)
+            if isinstance(spam_check_result, Codes):
                 raise SynapseError(
-                    403, "This message had been rejected as probable spam", spam_check
+                    403,
+                    "This message had been rejected as probable spam",
+                    spam_check_result,
+                )
+            elif (
+                isinstance(spam_check_result, str)
```

Comment:
having different behaviours for `str`s that are members of `Codes` and those that aren't feels very magical :-S

Reply:
Hopefully https://github.com/matrix-org/synapse/pull/12918/commits/3a71525a65bdbcdc6b2a408f0d5202ebc8e622c3 makes this clearer